testthat 3.1.0 (#537)
tidyverse/tidyverse.org: content/blog/testthat-3-1/index.Rmd

  author: Ethan Hoover

# one of: "deep-dive", "learn", "package", "programming", or "other"
categories: [package] 
tags: [testthat]
---

<!--
TODO:
* [x] Look over / edit the post's title in the yaml
* [x] Edit (or delete) the description; note this appears in the Twitter card
* [x] Pick category and tags (see existing with `hugodown::tidy_show_meta()`)
* [x] Find photo & update yaml metadata
* [x] Create `thumbnail-sq.jpg`; height and width should be equal
* [x] Create `thumbnail-wd.jpg`; width should be >5x height
* [x] `hugodown::use_tidy_thumbnails()`
* [x] Add intro sentence, e.g. the standard tagline for the package
* [x] `usethis::use_tidy_thanks()`
-->

We're stoked to announce the release of [testthat](http://testthat.r-lib.org/) 3.1.0.
testthat makes it easy to turn your existing informal tests into formal, automated tests that you can rerun quickly and easily.
testthat is the most popular unit-testing package for R, and is used by over 6,000 CRAN and Bioconductor packages.
You can learn more about unit testing at <https://r-pkgs.org/tests.html>.

You can install testthat from CRAN with:

```{r, eval = FALSE}
install.packages("{package}")
```

> This release of testthat includes a bunch of minor improvements to snapshotting as well as one breaking change (which only applies if you're using the 3rd edition).

Reviewer: ```suggestion
This release of testthat includes a bunch of minor improvements to snapshotting, as well as one breaking change (which only applies if you're using the 3rd edition).
```

You can see a full list of changes in the [release notes](https://github.com/r-lib/testthat/blob/master/NEWS.md).

```{r setup}
library(testthat)
```

## Snapshot tests

> Most of the effort in this release has gone in [snapshot tests](https://testthat.r-lib.org/articles/snapshotting.html), a new feature in testthat 3.0.0.

Reviewer: ```suggestion
Most of the effort in this release has gone into [snapshot tests](https://testthat.r-lib.org/articles/snapshotting.html), a new feature in testthat 3.0.0.
```

Reviewer: IDK why, but it sounds weird to me as "in" as opposed to "into."

Author: Yeah, agreed

> snapshot tests (also known as [golden tests](https://ro-che.info/articles/2017-12-04-golden-tests)) record expect output in a separate human-readable file Instead of using code to describe expected output.

Reviewer 2: ```suggestion
snapshot tests (also known as [golden tests](https://ro-che.info/articles/2017-12-04-golden-tests)) record expected output in a separate human-readable file instead of using code to describe what the expected output looks like.
```

Since the release of testthat 3.0.0, we've started using snapshot tests across a bunch of tidyverse packages and they've been working out really well.

> I don't anticipate any major changes (although we may continue to add new features) so the snapshot functions have changed lifecycle stages from experimental to **stable**.

Reviewer: ```suggestion
I don't anticipate any major changes (although we may continue to add new features), so the snapshot functions have changed lifecycle stages from experimental to **stable**.
```

This release also includes two new features that help you use snapshot tests in more places:

-   `expect_snapshot()` gains a `transform` argument, which should be a function that takes a character vector of lines and returns a modified character vector of lines.
    This makes it easy to remove sensitive (e.g. API keys) or stochastic (e.g. random temporary directory names) data from snapshot output.

    ```{r}
    get_info <- function() {
      list(
        name = "Hadley", 
        password = "sssh-its-a-secret"
      )
    }

    hide_password <- function(x) {
      is_password <- grepl("password", x)
      ifelse(is_password, "<REDACTED>", x)
    }

    test_that("info retruns name and password", {
      local_edition(3)
      expect_snapshot(str(get_info()), transform = hide_password)
    })
    ```

    If you need `transform`, I recommend designing your printing methods so the output can be easily manipulated with regexps.

> -   `expect_snapshot()` and friends gets an experimental new `variant` argument which causes the snapshot to be saved in `_snaps/{variant}/{test}.md` instead of `_snaps/{test}.md`.

Reviewer 2: ```suggestion
-   `expect_snapshot()` and friends get an experimental new `variant` argument which causes the snapshot to be saved in `_snaps/{variant}/{test}.md` instead of `_snaps/{test}.md`.
```

      expect_snapshot(version$nickname, variant = r_version())
    })
    ```

Remember that snapshot tests are not run on CRAN by default because they require a human to confirm whether or not a change is a breakage, so you shouldn't rely only on snapshot tests to ensure that your code is correct.
While it is possible to set `cran = TRUE`, to force snapshot tests to run on CRAN, I don't generally recommend it as snapshots are often vulnerable to minor changes that don't merit breaking your released package.

## Breaking changes

We made one breaking change that affects the [third edition](https://testthat.r-lib.org/articles/third-edition.html).
Previously, `expect_message()`, `expect_warning()` and `expect_error()` returned the value of the first argument, unless that was `NULL`, when they instead returned the condition object.
This meant you could write code like this:

```{r, eval = FALSE}
expect_equal(expect_warning(f(), "warning"), "value")
```

Now `expect_message()`, `expect_warning()` and `expect_error()` always return the condition object so you need to flip the order of the expectations:

```{r, eval = FALSE}
expect_warning(expect_equal(f(), "value"), "warning")
```

> This (IMO) is a little easier to read with the pipe:

Reviewer 2: oh good point

```{r, eval = FALSE}
# Or use the pipe:
f() %>% 
  expect_equal("value") %>% 
  expect_warning("warning")
```

Or with an intermediate object:

```{r, eval = FALSE}
expect_warning(value <- f(), "warning")
expect_equal(value, "value")
```

As with any breaking change, we made this change with great care.
Fortunately it only affects the 3rd edition, which relatively few packages use, and we submitted PRs to all affected packages on CRAN.

We made this change because it makes testthat more consistent and makes it easier to inspect both the value and the warning.
This is important because it makes it easier to test functions that produce [custom error objects](https://adv-r.hadley.nz/conditions.html#custom-conditions) that themselves contain meaningful data:

```{r, eval = FALSE}
library(rlang)
informative_error <- function() {
  abort(
    "An error with extra info",
    name = "patrice",
    number = 17
  )
}

err <- expect_error(my_function(), class = "package_error_class")
expect_equal(err$name, "patrice")
expect_equal(err$number, 17)
```

> Richer conditions are tool that we use within the tidyverse to provide more context in error message.

Reviewer 2: ```suggestion
Richer conditions are tool that we use within the tidyverse to provide more context in errors, warnings, and messages.
```

Since `expect_error()` always returned the condition and the change only concerns warnings and messages.

Reviewer: In addition to lionel's change, the first part should be either "Richer conditions are **a** tool" or "Richer conditions are **tools**"
